tools/wa_keyring.1
.\" Automatically generated by Pod::Man 2.27 (Pod::Simple 3.28)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
.    ds C`
.    ds C'
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.\"
.\" Avoid warning from groff about undefined register 'F'.
.de IX
..
.nr rF 0
.if \n(.g .if rF .nr rF 1
.if (\n(rF:(\n(.g==0)) \{
.    if \nF \{
.        de IX
.        tm Index:\\$1\t\\n%\t"\\$2"
..
.        if !\nF==2 \{
.            nr % 0
.            nr F 2
.        \}
.    \}
.\}
.rr rF
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "WA_KEYRING 1"
.TH WA_KEYRING 1 "2014-12-10" "4.7.0" "WebAuth"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
wa_keyring \- WebAuth keyring manipulation tool
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
\&\fBwa_keyring\fR [\fB\-hv\fR] \fB\-f\fR \fIfile\fR \fIcommand\fR [\fIarg\fR ...]
.PP
\&\fBwa_keyring\fR \fB\-f\fR \fIkeyring\fR add \fIvalid-after\fR
.PP
\&\fBwa_keyring\fR \fB\-f\fR \fIkeyring\fR gc \fIoldest-valid-after-to-keep\fR
.PP
\&\fBwa_keyring\fR \fB\-f\fR \fIkeyring\fR list
.PP
\&\fBwa_keyring\fR \fB\-f\fR \fIkeyring\fR remove \fIid\fR
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
\&\fBwa_keyring\fR is a command line tool to manage WebAuth key ring files,
which contain the private \s-1AES\s0 keys used by mod_webauth and mod_webkdc.  It
supports the following individual commands:
.IP "add \fIvalid-after\fR" 4
.IX Item "add valid-after"
Adds a new key to the key ring.  \fIvalid-after\fR uses the format:
.Sp
.Vb 1
\&    nnnn[s|m|h|d|w]
.Ve
.Sp
to indicate a time relative to the current time. The units for the time
are specified by appending a single letter.  That letter can be any of s,
m, h, d, or w, which correspond to seconds, minutes, hours, days, and
weeks respectively.
.Sp
For example: 10d is 10 days from the current time, and \-60d is 60 days
before the current time.
.IP "gc \fIoldest-valid-after-to-keep\fR" 4
.IX Item "gc oldest-valid-after-to-keep"
Garbage collects (removes) old keys on the key ring.  Any keys with a
\&\fIvalid-after\fR date older then the specified time will be removed from the
key ring.
.Sp
The format for \fIoldest-valid-after-to-keep\fR is the same as \fIvalid-after\fR
from the add command.  Note that this means that times given to the gc
command should generally be negative, to remove keys that have expired in
the past.
.IP "list" 4
.IX Item "list"
Lists all the keys in the key ring.  By default, a brief listing is
used, but a verbose listing can be requested with the \fB\-v\fR option.
.Sp
The following fields are present in a short listing:
.RS 4
.IP "\fBid\fR" 4
.IX Item "id"
The index/position of the key in the key ring.
.IP "\fBCreated\fR" 4
.IX Item "Created"
The date the key was created.
.IP "\fBValid after\fR" 4
.IX Item "Valid after"
The date at which the key becomes valid (in other words, the point at
which the WebAuth server will start using it to encrypt and decrypt new
data).
.IP "\fBFingerprint\fR" 4
.IX Item "Fingerprint"
The \s-1MD5\s0 digest of the key data.  Used to compare keys in two key rings.
.RE
.RS 4
.Sp
The following fields are present in the long listing:
.IP "\fBKey-Id\fR" 4
.IX Item "Key-Id"
The index/position of the key in the key ring.
.IP "\fBCreated\fR" 4
.IX Item "Created"
The date the key was created.
.IP "\fBValid-After\fR" 4
.IX Item "Valid-After"
The date at which the key becomes valid (in other words, the point at
which the WebAuth server will start using it to encrypt and decrypt new
data).
.IP "\fBKey-Type\fR" 4
.IX Item "Key-Type"
The type of key.  Currently, \s-1AES\s0 is the only supported key type.
.IP "\fBKey-Size\fR" 4
.IX Item "Key-Size"
Length in bytes of the key.
.IP "\fBFingerprint\fR" 4
.IX Item "Fingerprint"
The \s-1MD5\s0 digest of the key data. Used to compare keys in two key rings.
.RE
.RS 4
.RE
.IP "remove \fIid\fR" 4
.IX Item "remove id"
Remove the key with \s-1ID \s0\fIid\fR from the key ring.
.PP
For any of the commands that change the keyring, \fBwa_keyring\fR must have
write access to the directory containing the keyring, since keyrings are
updated by writing out the new file to a separate name and then atomically
replacing the file.
.PP
Ownership (user and group) of the existing keyring file will be preserved
if possible without overwriting the existing file.  Permissions will also
be preserved, with the exception that permissions will not be copied to
the new file if the old file was group-readable or group-writable and
setting the group ownership failed.
.SH "EXAMPLES"
.IX Header "EXAMPLES"
Add a key to the keyring valid as of the current time:
.PP
.Vb 1
\&    wa_keyring \-f keyring add 0d
.Ve
.PP
Add a key to the keyring that will be valid three days from now:
.PP
.Vb 1
\&    wa_keyring \-f keyring add 3d
.Ve
.PP
Remove keys from the key ring that became invalid more than 90 days ago:
.PP
.Vb 1
\&    wa_keyring \-f keyring gc \-90d
.Ve
.PP
Remove the first key in the keyring.
.PP
.Vb 1
\&    wa_keyring \-f keyring remove 0
.Ve
.PP
Display a verbose listing of all of the keys in the key ring:
.PP
.Vb 1
\&    wa_keyring \-f keyring \-v list
.Ve
.PP
Note that a WebAuth server will normally manage its keyring file by
itself, and \fBwa_keyring\fR is normally only used for debugging purposes.
However, if you are setting up a load-balanced pool of servers that need
to all share the same keys, turn off automatic keyring handling by putting
the line:
.PP
.Vb 1
\&    WebAuthKeyringAutoUpdate off
.Ve
.PP
to your Apache configuration, running a script periodically from cron on
one server that does something like:
.PP
.Vb 2
\&    wa_keyring \-f keyring gc \-90d
\&    wa_keyring \-f keyring add 2d
.Ve
.PP
and then copying (in a secure manner!) the new keyring file to all of the
other servers.
.SH "AUTHOR"
.IX Header "AUTHOR"
Roland Schemers <schemers@stanford.edu>
.SH "COPYRIGHT AND LICENSE"
.IX Header "COPYRIGHT AND LICENSE"
Copyright 2002, 2004, 2005, 2014 The Board of Trustees of the Leland
Stanford Junior University
.PP
Copying and distribution of this file, with or without modification, are
permitted in any medium without royalty provided the copyright notice and
this notice are preserved.  This file is offered as-is, without any
warranty.


NEWS
                       User-Visible WebAuth Changes

WebAuth 4.7.0 (2014-12-10)

    Recognize KRB5_BAD_ENCTYPE, KRB5_GET_IN_TKT_LOOP, KRB5_PREAUTH_FAILED,
    and KRB5KRB_AP_ERR_MODIFIED as additional synonyms for a failed login
    error code.  Various combinations of recent MIT and Heimdal with
    different KDCs return these error codes if the password is incorrect.

    Added new fields to the userinfo service parsing and the WebLogin
    handling.  These allow for a more complicated multifactor
    configuration to be passed along from the user information service,
    with multiple possible multifactor devices and one default.

    Give a validation remctl command its own timeout error, so that a
    failure to respond to validation is handled differently than any
    other timeout failure.  This is done so that we can handle
    out-of-band multifactor methods, such as a phone call.  Previously
    that would show up in WebLogin as a generic
    WK_ERR_UNRECOVERABLE_ERROR.

    Ability to use JSON rather than XML for the user information service's
    return values.  This is activated with the WebKdcUserInfoJSON
    configuration directive.

    Refactored the userinfo code to separate remctl support, XML parsing,
    and JSON parsing into separate source files for readability.

WebAuth 4.6.1 (2014-07-23)

    Support for AuthType StanfordAuth (for backward compatibility with
    WebAuth 2.5) was broken in WebAuth 4.6.0, causing mod_webauth to
    reject all accesses to resources protected with that AuthType.  This
    has been fixed in this release.

    Add a new configuration directive, WebKdcFastArmorCache, for
    mod_webkdc.  If set, this specifies the path to a Kerberos ticket
    cache that can (and must) be used for FAST (Flexible Authentication
    Secure Tunneling) protection of Kerberos password authentications.
    The Kerberos KDC must also support FAST in order to safely enable this
    option.  Based on a patch by Jakob Uhd Jepsen (One.com A/S).

    Fix parsing of the WebKdcKerberosFactors configuration directive.

    Add a new webauth_krb5_set_fast_armor_path interface to libwebauth
    that allows configuring a path to a FAST armor ticket cache before
    authenticating with a password.

    Show the expiring password warning in WebLogin if the browser request
    was a POST.  Previously, it was skipped if the user had a REMOTE_USER
    preference or if the browser presented a single sign-on cookie.  This
    was too conservative, not warning in cases when REMOTE_USER failed,
    when the browser presented an expired single sign-on cookie (systems
    that are suspended rather than shut down, for example), and when the
    user has to do multifactor authentication.  Checking for a POST is a
    closer match for when we can force a confirmation screen without too
    much user disruption.

    When translating Kerberos errors, treat KRB5_KDC_UNREACH (cannot
    contact any KDC for realm) as a user rejected error instead of a
    Kerberos error.  This avoids returning an internal error from WebLogin
    and instead tells the user the username is invalid.  This is not
    always correct, since the unreachable KDC could be the local KDC, but
    it's better than the previous behavior of throwing internal errors
    when users enter email addresses as their username.

    Translate an EINVAL error from the Kerberos libraries during password
    authentication to an incorrect password error code.  Older versions of
    MIT Kerberos returned EINVAL for excessively long passwords.

    In WebLogin, verify that the username form field was sent before
    attempting to do multifactor operations and return an error if it
    isn't, avoiding undefined variable warnings and other errors deeper in
    the WebLogin code.

    Allow newlines, carriage returns, and tabs in the XML sent from the
    WebKDC to the WebLogin server rather than replacing them with periods.
    This fixes the display of <user-message> elements that contain
    newlines.

    If a user may switch to a different authorization identity, force
    display of the confirmation page in WebLogin even if this is normally
    disabled.  Otherwise, there is no opportunity for the user to change
    identities.

    Diagnose empty RT or ST parameters to WebLogin and return the same
    error as when those parameters are missing entirely.

    Fix compilation when remctl support is not enabled.

    Add new factors mp (mobile push) and v (voice), which count as
    separate classes for determining multifactor.  This means the
    combination of those factors with any other factor class will result
    in a synthensized multifactor factor.

    Warn in the mod_webauth documentation that, when using credential
    delegation to a load-balanced pool, all members of that pool must have
    the same Kerberos identity.

    Update to rra-c-util 5.5:

    * Use Lancaster Consensus environment variables to control tests.
    * Use calloc or reallocarray for protection against integer overflows.
    * Suppress warnings from Kerberos headers in non-system paths.
    * Update warning flags when building with make warnings.
    * Only pass warning suppression flags to Perl under make warnings.

    Update to C TAP Harness 3.1:

    * Check for integer overflow on memory allocations.
    * Avoid all remaining uses of sprintf.

WebAuth 4.6.0 (2014-03-18)

    WARNING: When upgrading to this release, you will need to change the
    ownership of the mod_webauth keyring file (named in the WebAuthKeyring
    Apache directive) to the user and group the Apache child processes run
    as.  (This is controlled by the User and Group Apache directives and
    is often something like www-data.)  Previous versions would create
    this file readable only by root, but, as of this release, it must be
    readable and writable by the Apache child processes.

    mod_webauth supports a new configuration directive, WebAuthCookiePath,

tools/wa_keyring.c
/*
 * Command-line utility for manipulating WebAuth keyrings.
 *
 * Written by Roland Schemers and Russ Allbery
 * Copyright 2002, 2003, 2006, 2009, 2010, 2012
 *     The Board of Trustees of the Leland Stanford Junior University
 *
 * See LICENSE for licensing terms.
 */

#include <config.h>
#include <portable/apr.h>
#include <portable/stdbool.h>
#include <portable/system.h>

#include <assert.h>
#include <errno.h>
#include <openssl/md5.h>
#include <time.h>

#include <util/messages.h>
#include <util/xmalloc.h>
#include <webauth/basic.h>
#include <webauth/keys.h>

/* Usage message. */
static const char usage_message[] = "\
Usage: %s [-hv] -f <keyring> list\n\
       %s -f <keyring> add <valid-after>\n\
       %s -f <keyring> gc <oldest-valid-after-to-keep>\n\
       %s -f <keyring> remove <id>\n\
\n\
Functions:\n\
  add <valid-after>                 # add a new random key\n\
  gc <oldest-valid-after-to-keep>   # garbage collect old keys\n\
  list                              # list keys\n\
  remove <id>                       # remove key by id\n\
\n\
<valid_after> and <oldest-valid-after-to-keep> use the format\n\
[-]<nnnn>[s|m|h|d|w], indicating a time relative to the current time.  The\n\
units for the time are specified by appending a single letter, which can\n\
be s, m, h, d, or w, corresponding to seconds minutes, hours, days, and\n\
weeks, respectively.  For example, 10d is 10 days from the current time,\n\
and -60d is 60 days before the current time.  Negative relative times are\n\
useful with gc.\n";


/*
 * Die with an error, appending a WebAuth error message.  Tries to follow the
 * interface and behavior of the messages library as closely as possible.
 */
static void
die_webauth(struct webauth_context *ctx, int s, const char *fmt, ...)
{
    va_list args;

    if (message_program_name != NULL)
        fprintf(stderr, "%s: ", message_program_name);
    va_start(args, fmt);
    vfprintf(stderr, fmt, args);
    if (s != 0)
        fprintf(stderr, ": %s", webauth_error_message(ctx, s));
    fprintf(stderr, "\n");
    exit(message_fatal_cleanup ? (*message_fatal_cleanup)() : 1);
}


/*
 * Display the usage message for remctl.
 */
static void
usage(int status)
{
    fprintf((status == 0) ? stdout : stderr, usage_message,
            message_program_name, message_program_name, message_program_name,
            message_program_name);
    exit(status);
}


/*
 * Convert a time value to seconds, supporting a single character for units.
 * s, m, h, d, and w are the supported units, representing seconds, minutes,
 * hours, days, and weeks respectively.
 */
static long
seconds(const char *value)
{
    long n;
    long multiplier = 0;
    char *end;

    assert(value != NULL);
    if (*value == '\0')
        die("invalid empty time value");

    switch (value[strlen(value) - 1]) {
    case 's':
        multiplier = 1;
        break;
    case 'm':
        multiplier = 60;
        break;
    case 'h':
        multiplier = 60 * 60;
        break;
    case 'd':
        multiplier = 60 * 60 * 24;
        break;
    case 'w':
        multiplier = 60 * 60 * 24 * 7;
        break;
    default:
        die("invalid time specification: %s", value);
        break;
    }

    errno = 0;
    n = strtol(value, &end, 10);
    if (errno != 0 || end != value + strlen(value) - 1)
        die("invalid time specification: %s", value);
    return n * multiplier;
}


/*
 * Given a time_t, print it to standard output in the local time zone using
 * the ISO date and time representation but without the time zone.
 */
static void
print_time(time_t t)
{
    struct tm *tm;
    char buf[128];

    tm = localtime(&t);
    strftime(buf, sizeof(buf), "%Y-%m-%d %H:%M:%S", tm);
    printf("%s", buf);
}


/*
 * Given a WebAuth key, print a fingerprint of that key to standard output.
 * The fingerprint is the MD5 checksum of the key.  This is used to check
 * whether two keys are the same without exposing the actual key value.
 */
static void
print_fingerprint(struct webauth_key *key)
{
    unsigned char md5[MD5_DIGEST_LENGTH];
    size_t i;
    MD5_CTX c;

    MD5_Init(&c);
    MD5_Update(&c, key->data, key->length);
    MD5_Final(md5, &c);
    for (i = 0; i < sizeof(md5); i++)
        printf("%.2x", (unsigned int) md5[i]);
}


/*
 * Given a keyring entry, print it in the short format.  Each entry will look
 * like:
 *
 *     0  2009-06-27 15:29:19  2009-06-27 15:29:19  <fingerprint>
 *
 * where the first column is the id, the next is the creation timestamp, the
 * third is the date at which the key becomes valid, and the last is the MD5
 * fingerprint of the key.  Takes the keyring entry and its index, which is
 * used as the key ID.
 */
static void
print_short(struct webauth_keyring_entry *e, size_t i)
{
    printf("%2lu  ", (unsigned long) i);
    print_time(e->creation);
    printf("  ");
    print_time(e->valid_after);
    printf("  ");
    print_fingerprint(e->key);
    printf("\n");
}


/*
 * Given a keyring entry, print it in the long format.  Each entry will look
 * like:
 *
 *          Key-Id: 0
 *         Created: 2009-06-27 15:29:19
 *     Valid-After: 2009-06-27 15:29:19
 *        Key-Type: 1 (AES)
 *      Key-Length: 16 (128 bits)
 *     Fingerprint: 23e59c1d193113afc9d0257f42a64369
 *
 * Takes the keyring entry and its index, which is used as the key ID.
 */
static void
print_long(struct webauth_keyring_entry *e, size_t i)
{
    printf("       Key-Id: %lu\n", (unsigned long) i);
    printf("      Created: ");
    print_time(e->creation);
    printf("\n");
    printf("  Valid-After: ");
    print_time(e->valid_after);
    printf("\n");
    printf("     Key-Type: %d (", e->key->type);
    switch (e->key->type) {
    case WA_KEY_AES:
        printf("AES");
        break;
    default:
        printf("UNKNOWN");
        break;
    }
    printf(")\n");
    printf("   Key-Length: %d bits\n", e->key->length * 8);
    printf("  Fingerprint: ");
    print_fingerprint(e->key);
    printf("\n\n");
}


/*
 * List the entries in the keyring.  The default output format looks like:
 *
 *     Path: webauth_keyring
 *
 *     id  Created              Valid after          Fingerprint
 *      0  2009-06-27 15:29:19  2009-06-27 15:29:19  23e59c1d193113afc9...
 *
 * If the verbose flag is true, the output will instead look like:
 *
 *             Path: keyring
 *         Num-Keys: 1
 *
 *           Key-Id: 0
 *          Created: 2009-06-27 15:29:19
 *      Valid-After: 2009-06-27 15:29:19
 *         Key-Type: 1 (AES)
 *       Key-Length: 128 bits
 *      Fingerprint: 23e59c1d193113afc9d0257f42a64369
 *
 * Takes the path to the keyring and the verbose flag.
 */
static void
list_keyring(struct webauth_context *ctx, const char *keyring, bool verbose)
{
    struct webauth_keyring *ring;
    struct webauth_keyring_entry *entry;
    int s;
    size_t i;

    s = webauth_keyring_read(ctx, keyring, &ring);
    if (s != WA_ERR_NONE)
        die_webauth(ctx, s, "cannot read keyring %s", keyring);
    if (verbose) {
        printf("         Path: %s\n", keyring);
        printf("     Num-Keys: %d\n\n", ring->entries->nelts);
    } else {
        printf("Path: %s\n", keyring);
        printf("\n");
        printf("id  Created              Valid after          Fingerprint\n");
    }
    for (i = 0; i < (size_t) ring->entries->nelts; i++) {
        entry = &APR_ARRAY_IDX(ring->entries, i, struct webauth_keyring_entry);
        if (verbose)
            print_long(entry, i);
        else
            print_short(entry, i);
    }
}


/*
 * Add a new key to a keyring.  Takes the path to the keyring and the offset
 * in seconds at which the new key should become valid.
 */
static void
add_key(struct webauth_context *ctx, const char *keyring, long valid_after)
{
    struct webauth_key *key;
    struct webauth_keyring *ring;
    int s;
    time_t now;

    s = webauth_key_create(ctx, WA_KEY_AES, WA_AES_128, NULL, &key);
    if (s != WA_ERR_NONE)
        die_webauth(ctx, s, "cannot generate new random key");

    /* FIXME: Don't generate a new keyring on any error. */
    s = webauth_keyring_read(ctx, keyring, &ring);
    if (s != WA_ERR_NONE)
        ring = webauth_keyring_new(ctx, 1);
    now = time(NULL);
    webauth_keyring_add(ctx, ring, now, now + valid_after, key);
    s = webauth_keyring_write(ctx, ring, keyring);
    if (s != WA_ERR_NONE)
        die_webauth(ctx, s, "cannot write keyring to %s", keyring);
}


/*
 * Remove a key from a keyring.  Takes the index in the keyring of the key to
 * remove.
 */
static void
remove_key(struct webauth_context *ctx, const char *keyring, unsigned long n)
{
    struct webauth_keyring *ring;
    int s;

    s = webauth_keyring_read(ctx, keyring, &ring);
    if (s != WA_ERR_NONE)
        die_webauth(ctx, s, "cannot read keyring %s", keyring);
    s = webauth_keyring_remove(ctx, ring, n);
    if (s != WA_ERR_NONE)
        die_webauth(ctx, s, "cannot remove key %lu from keyring", n);
    s = webauth_keyring_write(ctx, ring, keyring);
    if (s != WA_ERR_NONE)
        die_webauth(ctx, s, "cannot write keyring to %s", keyring);
}


/*
 * Garbage-collect the keys in the keyring.  Any key whose valid-after date is
 * older than the current time adjusted by the offset passed in to gc_keys is
 * removed from the keyring.
 */
static void
gc_keys(struct webauth_context *ctx, const char *keyring, long offset)
{
    struct webauth_keyring *ring;
    struct webauth_keyring_entry *entry;
    int s;
    bool removed;
    size_t i;
    time_t now, earliest;

    s = webauth_keyring_read(ctx, keyring, &ring);
    if (s != WA_ERR_NONE)
        die_webauth(ctx, s, "cannot read keyring %s", keyring);
    now = time(NULL);
    earliest = now + offset;
    do {
        removed = false;
        for (i = 0; i < (size_t) ring->entries->nelts; i++) {
            entry = &APR_ARRAY_IDX(ring->entries, i,
                                   struct webauth_keyring_entry);
            if (entry->valid_after < earliest) {
                s = webauth_keyring_remove(ctx, ring, i);
                if (s != WA_ERR_NONE)
                    die_webauth(ctx, s, "cannot remove key %lu from keyring",
                                i);
                removed = true;
            }
        }
    } while (removed);
    s = webauth_keyring_write(ctx, ring, keyring);
    if (s != WA_ERR_NONE)
        die_webauth(ctx, s, "cannot write keyring to %s", keyring);
}


int
main(int argc, char **argv)
{
    int option, status;
    bool verbose = false;
    unsigned long id;
    long offset;
    char *end;
    const char *keyring = NULL;
    const char *command = "list";
    struct webauth_context *ctx;

    message_program_name = argv[0];

    /*
     * The + tells GNU getopt to stop option parsing at the first non-argument
     * rather than proceeding on to find options anywhere.  This allows easier
     * specification of negative time offsets.
     */
    while ((option = getopt(argc, argv, "+f:hv")) != EOF) {
        switch (option) {
        case 'f':
            keyring = optarg;
            break;
        case 'h':
            usage(0);
            break;
        case 'v':
            verbose = true;
            break;
        case '+':
            fprintf(stderr, "%s: invalid option -- +\n", argv[0]);
        default:
            usage(1);
            break;
        }
    }
    argc -= optind;
    argv += optind;
    if (keyring == NULL || argc > 2)
        usage(1);
    if (argc > 0) {
        command = argv[0];
        argc--;
        argv++;
    }

    status = webauth_context_init(&ctx, NULL);
    if (status != WA_ERR_NONE)
        die_webauth(NULL, status, "cannot initialize WebAuth");
    if (strcmp(command, "list") == 0) {
        if (argc > 0)
            usage(1);
        list_keyring(ctx, keyring, verbose);
    } else if (strcmp(command, "add") == 0) {
        if (argc != 1)
            usage(1);
        offset = seconds(argv[0]);
        add_key(ctx, keyring, offset);
    } else if (strcmp(command, "gc") == 0) {
        if (argc != 1)
            usage(1);
        offset = seconds(argv[0]);
        gc_keys(ctx, keyring, offset);
    } else if (strcmp(command, "remove") == 0) {
        if (argc != 1)
            usage(1);
        errno = 0;
        id = strtoul(argv[0], &end, 10);
        if (errno != 0 || *end != '\0')
            die("invalid key id: %s", argv[0]);
        remove_key(ctx, keyring, id);
    } else {
        usage(1);
    }
    webauth_context_free(ctx);
    exit(0);
}
